Explicitly allow non-numerical edges

scolobb opened the pull request:
The graph library does not really enforce edge weights to be numerical, but the documentation currently seems to imply that edge weights are protected by the contract `number?`. Apparently, this is not the case, as non-numerical edges can be created without any particular issues.

This pull request does the following modifications:

1. Modify the documentation to explicitly state that edge weights can be any values comparable with `equal?`.  EDIT: Explicitly say that some algorithms will not work for non-numerical weights.  (I think BFS and DFS should be fine, probably also SCCs, spanning trees, etc.)
2. Modify the `graphviz` function to compare edge weights with `equal?` instead of `=` (@stchang 's idea).
3. Add some new tests for `graphviz` to ensure that non-numerical edge labels are rendered correctly.
4. Fix a small mistake in some other tests of `graphviz` (essentially, escape or don't use square brackets, because this captures too many patterns).

stchang commented:
Thanks a lot! I think this is a good change.

Would you mind adding a caveat in the docs, with each weighted constructor, that some (many?) of the algorithms are not implemented with non-numerical weights in mind and may not work.

(Presumably, some of the algorithms could easily be parameterized with an extra comparison function, but I'll investigate that later.)

scolobb commented:
Thank you @stchang for your quick reply: you commented here even before I added a link to this PR to the issue you opened :-)

> Would you mind adding a caveat in the docs, with each weighted constructor, that some (many?) of the algorithms are not implemented with non-numerical weights in mind and may not work.

I think this is what I did.  I added the following text to each weighted constructor: "Weights can be any Racket values comparable with @racket[equal?]. Note that some algorithms will fail on non-numerical egde weights."

> (Presumably, some of the algorithms could easily be parameterized with an extra comparison function, but I'll investigate that later.)

I'd expect most algorithms to need more than that: they'd probably also need addition, negative values, maybe even multiplication for some.  In this case, the custom edge type would probably have to implement its own numerical interface, which should make it fine to use with the current version of the code without modifications.

stchang commented:
Ah sorry, should have read more closely :) Thanks!

And yes I think you're right that most algos will probably require more substantial changes to properly work with non-numeric weights.
